Setup test suite

Tijani-Dia opened the pull request:
I created a testapp by running wagtail start testapp.
I also added boilerplate files.

Stormheg reviewed (commented):
Thanks @Tijani-Dia 👍 

You are on the right track. I do have a few comments.

## 1. Test folder structure

I'm not keen on having a `tests` folder inside a `testapp` folder. I think it should be the other way around, like wagtail-factories for example: https://github.com/wagtail/wagtail-factories/tree/master/tests

## 2. Running tests

As for running tests, I think we should consider using [pytest](https://docs.pytest.org/en/6.2.x/) in combination with [tox](https://tox.readthedocs.io/en/latest/index.html) to run tests. Instead of Django's `manage.py test`.

Pytest is a testing library that is a bit more advanced than Django's default testing mechanism. It supports a ton of useful plugins as well.

Tox is a tool that can be used to test a lot of different Python / package combinations. Since we support a few different Wagtail / Django combinations[[1]] I think we should properly test these combinations. 

Take a look at wagtail-factories for inspiration on how pytest and tox are used: https://github.com/wagtail/wagtail-factories/blob/master/tox.ini

## Next steps: add linting 

In my opinion the next step (after merging this PR) would be to add linting with [flake8](https://flake8.pycqa.org/en/latest/) and [isort](https://pycqa.github.io/isort/). This ensures code in this package follows a code style and keeps things generally 'tidy'. We care about high-quality code.

Wagtail-factories uses flake8 and isort too: https://github.com/wagtail/wagtail-factories/blob/master/tox.ini#L38

[1]: https://docs.wagtail.io/en/stable/releases/upgrading.html#compatible-django-python-versions

--------

This may all be a bit overwhelming 😅 

Please feel free to reach out to me or other mentors on Slack if you have any questions or need guidance implementing these suggestions.

Tijani-Dia commented:
> Thanks @Tijani-Dia +1
> 
> You are on the right track. I do have a few comments.
> ## 1. Test folder structure
> 
> I'm not keen on having a `tests` folder inside a `testapp` folder. I think it should be the other way around, like wagtail-factories for example: https://github.com/wagtail/wagtail-factories/tree/master/tests
> ## 2. Running tests
> 
> As for running tests, I think we should consider using [pytest](https://docs.pytest.org/en/6.2.x/) in combination with [tox](https://tox.readthedocs.io/en/latest/index.html) to run tests. Instead of Django's `manage.py test`.
> 
> Pytest is a testing library that is a bit more advanced than Django's default testing mechanism. It supports a ton of useful plugins as well.
> 
> Tox is a tool that can be used to test a lot of different Python / package combinations. Since we support a few different Wagtail / Django combinations[[1](https://docs.wagtail.io/en/stable/releases/upgrading.html#compatible-django-python-versions)] I think we should properly test these combinations.
> 
> Take a look at wagtail-factories for inspiration on how pytest and tox are used: https://github.com/wagtail/wagtail-factories/blob/master/tox.ini
> ## Next steps: add linting
> 
> In my opinion the next step (after merging this PR) would be to add linting with [flake8](https://flake8.pycqa.org/en/latest/) and [isort](https://pycqa.github.io/isort/). This ensures code in this package follows a code style and keeps things generally 'tidy'. We care about high-quality code.
> 
> Wagtail-factories uses flake8 and isort too: https://github.com/wagtail/wagtail-factories/blob/master/tox.ini#L38
> 
> This may all be a bit overwhelming sweat_smile
> 
> Please feel free to reach out to me or other mentors on Slack if you have any questions or need guidance implementing these suggestions.

Thanks @Stormheg  for the helpful feedback! I agree with you on all you said above and I will take it into consideration.

We talked about linting in the first meeting and If I understood we are going to use black instead of flake8 but I'm not sure.

Tijani-Dia commented on setup.py:
```diff
@@ -0,0 +1,3 @@
+from setuptools import setup
+
+setup()
```
Yes, I really don't have experience on it.

allcaps replied:
The current balance between setup.py and setup.cfg is fine. The `cfg` is the new and future format. I think you need the setup.py for `pin install --editable` and to be backward compatible with some older stacks. But nothing to worry about now.

Tijani-Dia commented on testapp/Dockerfile:
```diff
@@ -0,0 +1,60 @@
+# Use an official Python runtime based on Debian 10 "buster" as a parent image.
```
This came when starting the testapp. It's a default file. I will remove it.

allcaps replied:
Yep, drop.

Tijani-Dia commented on .gitignore:
```diff
@@ -0,0 +1,66 @@
+*.egg-info/
+.installed.cfg
+*.egg
+
```
Sure!

Stormheg commented:
> We talked about linting in the first meeting and If I understood we are going to use black instead of flake8 but I'm not sure.

Black is great! I'm using black instead of flake8 in my own projects so I'd be keen to use black instead of flake8.

allcaps commented on testapp/testapp/settings.py:
```diff
@@ -0,0 +1,162 @@
+# Application definition
+
+INSTALLED_APPS = [
+    "liveblog",
```
Is `liveblog` a good name? Maybe it is bit confusing? `liveblog` is pretty similar to names you would encounter in the Wagtail Live package.

Maybe we can just move all content from liveblog and put it directly in the `testapp` folder. Then `INSTALLED_APPS` should have `'testapp',`.

Tijani-Dia replied:
Alright! Actually the problem I have is handling migrations and models. That's the main reason I haven't still added tests for `LivePageMixin`. 

How do we create models/tables for tests? I tried to use fixtures but I would prefer another approach.

Stormheg commented on tests/settings.py:
```diff
@@ -0,0 +1,79 @@
+
+DATABASES = {
+    "default": {
+        "ENGINE": "django.db.backends.sqlite3",
```
**Open question:** tests will use the sqlite3 backend. Is this okay?

allcaps replied:
Yes, fine. Easiest and we do not use any db specific things.

Stormheg commented:
As discussed during today's progress meeting I've contributed improvements to the test setup.

@allcaps can you review this and flag anything that doesn't seem right?

I've added black and isort linting. I'd prefer to have some sort of command to automatically fix linting issues. Would a Makefile be okay? (e.g. `make format`, `make lint`..)

Stormheg commented on tox.ini:
```diff
@@ -0,0 +1,30 @@
+[tox]
+skipsdist = True
+envlist = 
+    py38-django-wagtail{211,213}
```
I've kept the envlist constrained for now. I don't see added value in having a massive envlist in the current fase of early development.

I do intend to add different Django versions to the envlist sometime in the future.

Tijani-Dia commented:
> I've added black and isort linting.

I added flake8 because I found their logs useful except for the line too long one.

Do you mind keeping it?

Stormheg commented:
> Do you mind keeping it?

I don't mind at all, added it! 👍  Included a makefile as an easy shortcut to run all tools.

Waiting for your approval @allcaps

allcaps commented:
Black is a formatting tool and Flake coding style rules. They both have something to say about code style, but Flake has a much wider scope and Black fits within Flake rules.

So both is fine.

Concrete example. 'No unused variable' is a Flake rule (and a good one to be alerted about) bit black just doesn't care.

Flake and Black sometimes collide. Like with line length, the trick is to give Flake the same or longer length.

Did you consider isort? It is related and complements the Flake and Black.

Tijani-Dia commented:
Yes, I've been using all three. 
> Flake and Black sometimes collide. Like with line length, the trick is to give Flake the same or longer length.

Alright, black and isort collide too sometimes.

allcaps commented:
@Tijani-Dia isort conf can go into setup.cfg and look like:

```
[isort]
; Be black compatible
multi_line_output=3
include_trailing_comma=True
force_grid_wrap=0
use_parentheses=True
line_length=88

; Project
default_section = THIRDPARTY
known_first_party = testapp
known_django = django
known_wagtail = wagtail
sections = FUTURE,STDLIB,THIRDPARTY,DJANGO,WAGTAIL,FIRSTPARTY,LOCALFOLDER
skip = migrations/*
```

Stormheg commented:
Oops, missed your comment @allcaps. Let's create a follow up PR for the isort config.
